lib/shoehorn/application_controller.ex

```diff
@@ -135,23 +153,29 @@ defmodule Shoehorn.ApplicationController do
     :application
     """)
 
-    s
+    {:ok, s}
   end
 
-  defp shutdown(app, s) do
-    shoehorn = self()
+  defp shutdown(app, message, s=%{status: level}) do
+    {:ok, shutdown_timer_ref} = :timer.apply_after(s.shutdown_timer, :erlang, :halt, [])
 
-    # start the shutdown timer
-    Process.send_after(shoehorn, :shutdown, s.shutdown_timer)
+    {action, new_handler_state} = apply(s.handler, :handle_application, [message, app, s.handler_state])
 
-    spawn(fn ->
-      apply(s.handler, :application_stopped, [app])
-      send(shoehorn, :shutdown)
-    end)
+    case action do
+      :halt ->
+        :erlang.halt()
+      :restart ->
+        Shoehorn.ApplicationController.ensure_all_started(app)
```

Comment:
The GenServer can't call itself here, I think this line needs to be removed.

Reply:
Fixed and squashed, sorry about that.